### PR TITLE
TY: Introduced TyProjection (replacement for associated type)

### PR DESCRIPTION
**Pull request by vlad20012**

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/types/ty/TyProjection.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * ```
+ * trait Trait { type Item; }
+ * fn foo<T: Trait>() -> T::Item { unimplemented!() }
+ * fn bar<T: Trait>() -> <T as Trait>::Item { unimplemented!() }
+ * ```
+ *
+ * Fields meaning:
+ * ```
+ * <T as Trait>::Item
+ * |     |       |
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/types/ty/TyProjection.kt]: I think we need extra space before first `|`

File: src/main/kotlin/org/rust/lang/core/types/BoundElement.kt [hunk 2]
```diff
@@ -7,36 +7,38 @@ package org.rust.lang.core.types
...
  * `fn make_t<T>() -> T { ... }`, together with actual
  * type arguments, like `T := i32`.
  */
 data class BoundElement<out E : RsElement>(
     val element: E,
-    val subst: Substitution = emptySubstitution
+    /** Generic type bindings, e.g. `type T = Box<u8>;` */
+    val subst: Substitution = emptySubstitution,
+    /** Associated type bindings, e.g. `type T = Iterator<Item=u8>;` */
+    val assoc: Map<RsTypeAlias, Ty> = emptyMap()
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/types/BoundElement.kt]: Maybe slightly change docs for these fields? It's a little bit confusing and I spent some time to understand it.
Maybe something like this:
"Generic type bindings, e.g. if we have concrete type `Box<u8>` with a generic declaration as `Box<T>`, `subst` maps `T` to `u8`"? 
And something similar for `assoc` map as well

File: src/main/kotlin/org/rust/lang/core/types/BoundElement.kt [hunk 3]
```diff
@@ -7,36 +7,38 @@ package org.rust.lang.core.types
...
 import org.rust.lang.core.types.ty.emptySubstitution
 import org.rust.lang.core.types.ty.foldValues
-import org.rust.lang.core.types.ty.substituteInValues
 
-/* Represents a potentially generic Psi Element, like
+/**
+ * Represents a potentially generic Psi Element, like
  * `fn make_t<T>() -> T { ... }`, together with actual
  * type arguments, like `T := i32`.
  */
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/types/BoundElement.kt]: Let's add some info about associated types too

**Comment by vlad20012** [comment]
Done :)

**Pull request merged**